torchvision/prototype/models/depth/stereo/crestereo.py

```diff
@@ -0,0 +1,1096 @@
+    # and the i-th candidate on the right feature map
+    return correlation
+
+
+def _check_window_specs(
+    search_window_1d: Tuple[int, int] = (1, 9),
+    search_dilate_1d: Tuple[int, int] = (1, 1),
+    search_window_2d: Tuple[int, int] = (3, 3),
+    search_dilate_2d: Tuple[int, int] = (1, 1),
+) -> None:
+
+    if not np.prod(search_window_1d) == np.prod(search_window_2d):
+        raise ValueError(
+            f"The 1D and 2D windows should contain the same number of elements. "
+            f"1D shape: {search_window_1d} 2D shape: {search_window_2d}"
+        )
+    if not np.prod(search_window_1d) % 2 == 1:
+        raise ValueError(
+            f"Search windows should contain an odd number of elements in them."
+            f"Window of shape {search_window_1d} has {np.prod(search_window_1d)} elements."
+        )
+    if not any(size == 1 for size in search_window_1d):
+        raise ValueError(f"The 1D search window should have at least one size equal to 1. 1D shape: {search_window_1d}")
+    if any(size == 1 for size in search_window_2d):
+        raise ValueError(
+            f"The 2D search window should have all dimensions greater than 1. 2D shape: {search_window_2d}"
+        )
+    if any(dilate < 1 for dilate in search_dilate_1d):
+        raise ValueError(
+            f"The 1D search dilation should have all elements equal or greater than 1. 1D shape: {search_dilate_1d}"
+        )
+    if any(dilate < 1 for dilate in search_dilate_2d):
+        raise ValueError(
+            f"The 2D search dilation should have all elements equal greater than 1. 2D shape: {search_dilate_2d}"
+        )
+
+
+class IterativeCorrelationLayer(nn.Module):
+    def __init__(
+        self,
+        groups: int = 4,
+        search_window_1d: Tuple[int, int] = (1, 9),
+        search_dilate_1d: Tuple[int, int] = (1, 1),
+        search_window_2d: Tuple[int, int] = (3, 3),
+        search_dilate_2d: Tuple[int, int] = (1, 1),
+    ) -> None:
+
+        super().__init__()
+        _check_window_specs(
+            search_window_1d=search_window_1d,
+            search_dilate_1d=search_dilate_1d,
+            search_window_2d=search_window_2d,
+            search_dilate_2d=search_dilate_2d,
+        )
+        self.search_pixels = np.prod(search_window_1d)
+        self.groups = groups
+
+        # two selection tables for dealing withh the small_patch argument in the forward function
+        self.patch_sizes = {
+            "2d": [search_window_2d for _ in range(self.groups)],
+            "1d": [search_window_1d for _ in range(self.groups)],
+        }
+
+        self.dilate_sizes = {
+            "2d": [search_dilate_2d for _ in range(self.groups)],
+            "1d": [search_dilate_1d for _ in range(self.groups)],
+        }
+
+    def _make_coords(self, feature_map: Tensor) -> Tensor:
+        return make_coords_grid(feature_map.shape[0], feature_map.shape[2], feature_map.shape[3]).to(feature_map.device)
```

Comment:
NIT, In this PR you introduce `device` param on `make_coords_grid`, I think in that case we should use it instead of using `.to(device)`. Also since this is only called once in this Class, maybe we dont need to have `_make_coords` and call the `make_coords_grid` directly